### PR TITLE
OpenAI: /v1/models and /v1/models/{model} compatibility

### LINKED ISSUE #2430: `/v1/models` OpenAI compatibility API
It's great to have OpenAI compatibility API! 
Unfortunately many applications require additional endpoints to work properly.
Namely `v1/models` (https://platform.openai.com/docs/api-reference/models) with list of available models is the one most frequently missing.

### PR DESCRIPTION
**Pull request by royjhan**
This PR adds compatibility with the /v1/models and /v1/models/{model} endpoint for listing models.

E.g.

`curl http://localhost:11434/v1/models`

```
{
  "object": "list",
  "data": [
    {
      "id": "mario:latest",
      "object": "model",
      "created": 1718141294,
      "owned_by": "ollama"
    },
    {
      "id": "nomic-embed-text:latest",
      "object": "model",
      "created": 1718054969,
      "owned_by": "ollama"
    },
    {
      "id": "llava:latest",
      "object": "model",
      "created": 1718049682,
      "owned_by": "ollama"
    },
    {
      "id": "mistral:latest",
      "object": "model",
      "created": 1717609491,
      "owned_by": "ollama"
    },
    {
      "id": "llama3:latest",
      "object": "model",
      "created": 1717451603,
      "owned_by": "ollama"
    }
  ]
}
```

Resolves #2430 
Resolves #2476 

Includes #5028

### CODE REVIEW

**Review by JerrettDavis — APPROVED** [review]
Looks good for the list endpoint. 

We should probably open issues for the remaining models endpoints if they aren't already tracked:

```
GET https://api.openai.com/v1/models/{model}
DELETE https://api.openai.com/v1/models/{model}
```

**Comment by sammcj** [comment]
This would be nice!

File: server/routes_test.go [hunk 1]
```diff
@@ -126,6 +127,25 @@ func Test_Routes(t *testing.T) {
 				assert.Equal(t, "test-model:latest", modelList.Models[0].Name)
 			},
 		},
+		{
+			Name:   "List Model Handler OpenAI",
```
Comment (jmorganca) [inline server/routes_test.go:131]: Let's add a few more tests on things where we've seen some issues, such as an empty list etc

File: server/routes.go [hunk 2]
```diff
@@ -987,6 +987,7 @@ func (s *Server) GenerateRoutes() http.Handler {
 
 	// Compatibility endpoints
 	r.POST("/v1/chat/completions", openai.Middleware(), s.ChatHandler)
```
Comment (jmorganca) [inline server/routes.go:989]: Doesn't have to be part of this change but we should name this `ChatMiddleware`

File: server/routes_test.go [hunk 3]
```diff
@@ -104,6 +105,24 @@ func Test_Routes(t *testing.T) {
 				assert.Empty(t, len(modelList.Models))
 			},
 		},
+		{
+			Name:   "List Model Handler OpenAI (no tags)",
```
Comment (jmorganca) [inline server/routes_test.go:109]: For test names a better format would be lowercase:

```
openai list models no tags
```

Comment (royjhan) [reply server/routes_test.go]: Do we want to make all the test names lowercase then?

File: docs/openai.md [hunk 4]
```diff
@@ -27,6 +27,8 @@ chat_completion = client.chat.completions.create(
     ],
     model='llama3',
 )
+
```
Comment (jmorganca) [inline docs/openai.md:30]: Awesome work adding the docs! Let's make this a separate PR we can merge once the latest Ollama release includes it

File: openai/openai.go [hunk 5]
```diff
@@ -270,7 +308,33 @@ func (w *writer) writeResponse(data []byte) (int, error) {
...
+}
+
+func (w *ListWriter) writeResponse(data []byte) (int, error) {
+	var listResponse api.ListResponse
+	err := json.Unmarshal(data, &listResponse)
+	if err != nil {
+		return 0, err
+	}
+
+	// list completion
```
Comment (jmorganca) [inline openai/openai.go:327]: is this comment important?

File: server/routes_test.go [hunk 6]
```diff
@@ -127,6 +146,25 @@ func Test_Routes(t *testing.T) {
 				assert.Equal(t, "test-model:latest", modelList.Models[0].Name)
 			},
 		},
+		{
+			Name:   "openai list models yes tags",
```
Comment (jmorganca) [inline server/routes_test.go:150]:
```suggestion
			Name:   "openai list models with tags",
```

File: server/routes_test.go [hunk 7]
```diff
@@ -104,6 +105,24 @@ func Test_Routes(t *testing.T) {
 				assert.Empty(t, len(modelList.Models))
 			},
 		},
+		{
+			Name:   "openai list models no tags",
```
Comment (jmorganca) [inline server/routes_test.go:109]: Perhaps an empty list test?

**Review by jmorganca — COMMENTED** [review]
Getting close, a few small comments

File: server/routes_test.go (hunk 7 above)
Comment (royjhan) [reply server/routes_test.go:109]: This test is an empty list test, can rename

File: server/routes_test.go [hunk 8]
```diff
@@ -105,6 +106,24 @@ func Test_Routes(t *testing.T) {
 				assert.Empty(t, len(modelList.Models))
 			},
 		},
+		{
+			Name:   "openai empty list test",
```
Comment (jmorganca) [inline server/routes_test.go:110]:
```suggestion
			Name:   "openai empty list",
```

File: openai/openai_test.go [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+				if err := json.NewDecoder(resp.Body).Decode(&chatResp); err != nil {
+					t.Fatal(err)
+				}
+
+				assert.Equal(t, "chat.completion", chatResp.Object)
+				assert.Equal(t, "Hello!", chatResp.Choices[0].Message.Content)
+			},
+		},
+		{
+			Name:       "OpenAI List Handler",
```
Comment (jmorganca) [inline openai/openai_test.go:65]: we don't need `OpenAI ` in here since it's already in `openai_test.go`, just `list handler` is great

File: openai/openai_test.go [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+)
+
+func TestMiddleware(t *testing.T) {
+	type testCase struct {
+		Name       string
+		Method     string
+		Path       string
+		TestPath   string
+		Middleware func() gin.HandlerFunc
+		Endpoint   func(c *gin.Context)
```
Comment (jmorganca) [inline openai/openai_test.go:23]: `Handler` is a better name for this

File: server/routes_test.go (hunk 3 above)
Comment (jmorganca) [reply server/routes_test.go]: yes although no need to change them all here

**Comment by jmorganca** [comment]
This one's close.. just a few comments around the tests

**Review by jmorganca — APPROVED** [review]
